Fix DDP issue where parameters share same grad_accumulator

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46755 Fix DDP issue where parameters share same grad_accumulator**

Closes  https://github.com/pytorch/pytorch/issues/41324
As reported in https://github.com/pytorch/pytorch/issues/41324, there is a bug in DDP when `find_unused_parameters=True` and 2 or more parameters share the same gradient accumulator.

In the reducer, we currently keep a mapping of grad accumulator to index and populate it with map[accumulator] = index, but this overwrites indices when the accumulator is the same. To fix this, switch the mapping values to a vector of indices to hold all such indices that share the same accumulator.

Differential Revision: [D24497388](https://our.internmc.facebook.com/intern/diff/D24497388/)

rohan-varma committed:
Fix DDP issue where parameters share same grad_accumulator

As reported in https://github.com/pytorch/pytorch/issues/41324, there is a bug in DDP when `find_unused_parameters=True` and 2 or more parameters share the same gradient accumulator.

In the reducer, we currently keep a mapping of grad accumulator to index and populate it with map[accumulator] = index, but this overwrites indices when the accumulator is the same. To fix this, switch the mapping values to a vector of indices to hold all such indices that share the same accumulator.

Differential Revision: [D24497388](https://our.internmc.facebook.com/intern/diff/D24497388/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix DDP issue where parameters share same grad_accumulator"


As reported in https://github.com/pytorch/pytorch/issues/41324, there is a bug in DDP when `find_unused_parameters=True` and 2 or more parameters share the same gradient accumulator.

In the reducer, we currently keep a mapping of grad accumulator to index and populate it with map[accumulator] = index, but this overwrites indices when the accumulator is the same. To fix this, switch the mapping values to a vector of indices to hold all such indices that share the same accumulator.

Differential Revision: [D24497388](https://our.internmc.facebook.com/intern/diff/D24497388/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -120,7 +120,16 @@ Reducer::Reducer(
         // to check for parameters for which no gradient is computed, if
         // find_unused_parameters=True.
         if (find_unused_parameters_) {
```
Maybe expand the comments here to include how we handle multiple parameters.

pritamdamania87 reviewed (approved):
Thanks for fixing this quickly!

rohan-varma committed:
Update on "Fix DDP issue where parameters share same grad_accumulator"


Closes  https://github.com/pytorch/pytorch/issues/41324
As reported in https://github.com/pytorch/pytorch/issues/41324, there is a bug in DDP when `find_unused_parameters=True` and 2 or more parameters share the same gradient accumulator.

In the reducer, we currently keep a mapping of grad accumulator to index and populate it with map[accumulator] = index, but this overwrites indices when the accumulator is the same. To fix this, switch the mapping values to a vector of indices to hold all such indices that share the same accumulator.

Differential Revision: [D24497388](https://our.internmc.facebook.com/intern/diff/D24497388/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix DDP issue where parameters share same grad_accumulator"


Closes  https://github.com/pytorch/pytorch/issues/41324
As reported in https://github.com/pytorch/pytorch/issues/41324, there is a bug in DDP when `find_unused_parameters=True` and 2 or more parameters share the same gradient accumulator.

In the reducer, we currently keep a mapping of grad accumulator to index and populate it with map[accumulator] = index, but this overwrites indices when the accumulator is the same. To fix this, switch the mapping values to a vector of indices to hold all such indices that share the same accumulator.

Differential Revision: [D24497388](https://our.internmc.facebook.com/intern/diff/D24497388/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix DDP issue where parameters share same grad_accumulator"


Closes  https://github.com/pytorch/pytorch/issues/41324
As reported in https://github.com/pytorch/pytorch/issues/41324, there is a bug in DDP when `find_unused_parameters=True` and 2 or more parameters share the same gradient accumulator.

In the reducer, we currently keep a mapping of grad accumulator to index and populate it with map[accumulator] = index, but this overwrites indices when the accumulator is the same. To fix this, switch the mapping values to a vector of indices to hold all such indices that share the same accumulator.

Differential Revision: [D24497388](https://our.internmc.facebook.com/intern/diff/D24497388/)

[ghstack-poisoned]
